core/templates/dev/head/pages/exploration_editor/ExplorationSaveService.js

```diff
@@ -0,0 +1,516 @@
+                    return;
+                  }
+
+                  // Record any fields that have changed.
+                  var metadataList = [];
+                  if (explorationTitleService.hasChanged()) {
+                    metadataList.push('title');
+                  }
+                  if (explorationObjectiveService.hasChanged()) {
+                    metadataList.push('objective');
+                  }
+                  if (explorationCategoryService.hasChanged()) {
+                    metadataList.push('category');
+                  }
+                  if (explorationLanguageCodeService.hasChanged()) {
+                    metadataList.push('language');
+                  }
+                  if (explorationTagsService.hasChanged()) {
+                    metadataList.push('tags');
+                  }
+
+                  // Save all the displayed values.
+                  explorationTitleService.saveDisplayedValue();
+                  explorationObjectiveService.saveDisplayedValue();
+                  explorationCategoryService.saveDisplayedValue();
+                  explorationLanguageCodeService.saveDisplayedValue();
+                  explorationTagsService.saveDisplayedValue();
+
+                  // TODO(sll): Get rid of the $timeout here.
+                  // It's currently used because there is a race condition: the
+                  // saveDisplayedValue() calls above result in autosave calls.
+                  // These race with the discardDraft() call that
+                  // will be called when the draft changes entered here
+                  // are properly saved to the backend.
+                  $timeout(function() {
+                    $modalInstance.close(metadataList);
+                  }, 500);
+                };
+
+                $scope.cancel = function() {
+                  explorationTitleService.restoreFromMemento();
+                  explorationObjectiveService.restoreFromMemento();
+                  explorationCategoryService.restoreFromMemento();
+                  explorationLanguageCodeService.restoreFromMemento();
+                  explorationTagsService.restoreFromMemento();
+
+                  $modalInstance.dismiss('cancel');
+                  alertsService.clearWarnings();
+                };
+              }
+            ]
+          }).result.then(function(metadataList) {
+            if (metadataList.length > 0) {
+              var commitMessage = (
+                'Add metadata: ' + metadataList.join(', ') + '.');
+              saveDraftToBackend(commitMessage,
+                openPublishExplorationModal);
+            } else {
+              openPublishExplorationModal();
+            }
+          });
+        } else {
+          // No further metadata is needed. Open the publish modal immediately.
+          openPublishExplorationModal();
+        }
+      },
+
+      getPublishExplorationButtonTooltip: function() {
```

Comment:
This and the following method are specific to just the exploration editor navbar UI. They shouldn't be here.